runtime/collections/rangemaps.cpp

```diff
@@ -0,0 +1,282 @@
+    range *ptr = (range *)koreAlloc(sizeof(range));
+    ptr->h = range_header();
+    ptr->start = val.value().first.start();
+    ptr->end = val.value().first.end();
+    return (SortRange)ptr;
+  } else {
+    KLLVM_HOOK_INVALID_ARGUMENT("Key not found for range map lookup");
+  }
+}
+
+rangemap hook_RANGEMAP_update(
+    SortRangeMap m, SortKItem start, SortKItem end, SortKItem value) {
+  return m->inserted(rng_map::Range<KElem>(start, end), value);
+}
+
+rangemap
+hook_RANGEMAP_updateRng(SortRangeMap m, SortKItem rng, SortKItem value) {
+  range *ptr = (range *)rng;
+  return hook_RANGEMAP_update(m, ptr->start, ptr->end, value);
+}
+
+rangemap hook_RANGEMAP_remove(SortRangeMap m, SortKItem start, SortKItem end) {
+  return m->deleted(rng_map::Range<KElem>(start, end));
+}
+
+rangemap hook_RANGEMAP_removeRng(SortRangeMap m, SortKItem rng) {
+  range *ptr = (range *)rng;
+  return hook_RANGEMAP_remove(m, ptr->start, ptr->end);
+}
+
+rangemap hook_RANGEMAP_difference(SortRangeMap m1, SortRangeMap m2) {
+  return m1->difference(*m2);
+}
+
+set hook_SET_unit(void);
+set hook_SET_element(block *);
+set hook_SET_concat(set *, set *);
+
+set hook_RANGEMAP_keys(SortRangeMap m) {
+  auto tmp = hook_SET_unit();
+  for (auto iter = rng_map::ConstRangeMapIterator<KElem, KElem>(*m);
+       iter.has_next(); ++iter) {
+    range *ptr = (range *)koreAlloc(sizeof(range));
+    ptr->h = range_header();
+    ptr->start = iter->first.start();
+    ptr->end = iter->first.end();
+    auto elem = hook_SET_element((SortRange)ptr);
+    tmp = hook_SET_concat(&tmp, &elem);
+  }
+  return tmp;
+}
+
+list hook_RANGEMAP_keys_list(SortRangeMap m) {
+  auto tmp = list().transient();
+  for (auto iter = rng_map::ConstRangeMapIterator<KElem, KElem>(*m);
+       iter.has_next(); ++iter) {
+    range *ptr = (range *)koreAlloc(sizeof(range));
+    ptr->h = range_header();
+    ptr->start = iter->first.start();
+    ptr->end = iter->first.end();
+    tmp.push_back((SortRange)ptr);
+  }
+  return tmp.persistent();
+}
+
+bool hook_RANGEMAP_in_keys(SortKItem key, SortRangeMap m) {
+  return m->contains(key);
+}
+
+list hook_RANGEMAP_values(SortRangeMap m) {
+  auto tmp = list().transient();
+  for (auto iter = rng_map::ConstRangeMapIterator<KElem, KElem>(*m);
+       iter.has_next(); ++iter) {
+    tmp.push_back(iter->second);
+  }
+  return tmp.persistent();
+}
+
+SortKItem hook_RANGEMAP_choice(SortRangeMap m) {
+  if (m->empty()) {
+    KLLVM_HOOK_INVALID_ARGUMENT("Cannot choose from an empty range map");
+  }
+  block *elem = m->treemap().root_data().first.start();
+  return (SortKItem)elem;
+}
+
+SortKItem hook_RANGEMAP_choiceRng(SortRangeMap m) {
+  if (m->empty()) {
+    KLLVM_HOOK_INVALID_ARGUMENT("Cannot choose from an empty range map");
+  }
+  auto pair = m->treemap().root_data();
+  range *ptr = (range *)koreAlloc(sizeof(range));
+  ptr->h = range_header();
+  ptr->start = pair.first.start();
+  ptr->end = pair.first.end();
+  return (SortRange)ptr;
```

Comment:
here's another inconsistency. You declared the return type of this function as KItem, but you're returning a term of sort Range. Again, either an injection needs to be wrapped around the term, or else the return type needs to be changed to SortRange.

Reply:
Returning sort `Range`.